File: mobile/lib/models/tower_state.dart
import 'package:firebase_database/firebase_database.dart';

/// Data model representing the full state of a single hydroponic tower,
/// as stored in Firebase Realtime Database under
/// `farms/farm_id_1/towers/tower_1`.
class TowerState {
  final int pumpSpeed; // 0–255
  final String pumpMode; // 'auto' | 'manual'
  final bool pumpState; // actual pump on/off
  final bool pumpSwitch; // operator intent (manual mode)
  final int intervalOnMin;
  final int intervalOffMin;
  final int moisture; // ADC raw value 0–4095
  final bool waterLevelLow;

  const TowerState({
    required this.pumpSpeed,
    required this.pumpMode,
    required this.pumpState,
    required this.pumpSwitch,
    required this.intervalOnMin,
    required this.intervalOffMin,
    required this.moisture,
    required this.waterLevelLow,
  });

  // ---------------------------------------------------------------------------
  // Factories
  // ---------------------------------------------------------------------------

  /// Parses a [DataSnapshot] from Firebase into a [TowerState].
  /// Delegates to [fromMap] so the parsing logic is testable without Firebase.
  factory TowerState.fromSnapshot(DataSnapshot snapshot) {
    final raw = Map<String, dynamic>.from(snapshot.value as Map);
    return TowerState.fromMap(raw);
  }

  /// Parses a plain [Map] (matching the Firebase schema) into a [TowerState].
  /// All fields use null-coalescing defaults so the app never crashes on
  /// partial data.
  factory TowerState.fromMap(Map<String, dynamic> data) {
    final sensors = data['sensors'] != null
        ? Map<String, dynamic>.from(data['sensors'] as Map)
        : <String, dynamic>{};

    return TowerState(
      pumpSpeed: (data['pump_speed'] as int?) ?? 0,
      pumpMode: (data['pump_mode'] as String?) ?? 'manual',
      pumpState: (data['pump_state'] as bool?) ?? false,
      pumpSwitch: (data['pump_switch'] as bool?) ?? false,
      intervalOnMin: (data['interval_on_min'] as int?) ?? 1,
      intervalOffMin: (data['interval_off_min'] as int?) ?? 1,
      moisture: (sensors['moisture'] as int?) ?? 0,
      waterLevelLow: (sensors['water_level_low'] as bool?) ?? false,
    );
  }

  // ---------------------------------------------------------------------------
  // Serialisation
  // ---------------------------------------------------------------------------

  /// Serialises this [TowerState] to a [Map] that matches the Firebase schema,
  /// including the nested `sensors` key.  Used for test round-trips.
  Map<String, dynamic> toMap() {
    return {
      'pump_speed': pumpSpeed,
      'pump_mode': pumpMode,
      'pump_state': pumpState,
      'pump_switch': pumpSwitch,
      'interval_on_min': intervalOnMin,
      'interval_off_min': intervalOffMin,
      'sensors': {
        'moisture': moisture,
        'water_level_low': waterLevelLow,
      },
    };
  }

  // ---------------------------------------------------------------------------
  // Computed properties
  // ---------------------------------------------------------------------------

  /// Maps [pumpSpeed] (0–255) to a 0.0–1.0 slider value.
  double get speedPercent => pumpSpeed / 255.0;

  /// Maps a 0.0–1.0 slider percentage to an integer in the range 0–255.
  static int sliderToSpeed(double percent) =>
      (percent * 255).round().clamp(0, 255);

  /// Returns `'Low'` when [waterLevelLow] is `true`, otherwise `'Good'`.
  String get waterLevelDisplay => waterLevelLow ? 'Low' : 'Good';

  // ---------------------------------------------------------------------------
  // Equality & hashing
  // ---------------------------------------------------------------------------

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is TowerState &&
        other.pumpSpeed == pumpSpeed &&
        other.pumpMode == pumpMode &&
        other.pumpState == pumpState &&
        other.pumpSwitch == pumpSwitch &&
        other.intervalOnMin == intervalOnMin &&
        other.intervalOffMin == intervalOffMin &&
        other.moisture == moisture &&
        other.waterLevelLow == waterLevelLow;
  }

  @override
  int get hashCode => Object.hash(
        pumpSpeed,
        pumpMode,
        pumpState,
        pumpSwitch,
        intervalOnMin,
        intervalOffMin,
        moisture,
        waterLevelLow,
      );

  @override
  String toString() => 'TowerState('
      'pumpSpeed: $pumpSpeed, '
      'pumpMode: $pumpMode, '
      'pumpState: $pumpState, '
      'pumpSwitch: $pumpSwitch, '
      'intervalOnMin: $intervalOnMin, '
      'intervalOffMin: $intervalOffMin, '
      'moisture: $moisture, '
      'waterLevelLow: $waterLevelLow)';
}

File: web/src/data/towerState.ts
// Port of the Flutter `TowerState` model (mobile/lib/models/tower_state.dart).
// Mirrors the Firebase Realtime Database schema at
// `farms/farm_id_1/towers/tower_1`. All parsing uses defaults so the app never
// crashes on partial data.

export type PumpMode = 'auto' | 'manual';

export interface TowerState {
  pumpSpeed: number; // 0–255
  pumpMode: PumpMode; // 'auto' | 'manual'
  pumpState: boolean; // actual pump on/off (firmware-owned)
  pumpSwitch: boolean; // operator intent (manual mode)
  intervalOnMin: number;
  intervalOffMin: number;
  moisture: number; // ADC raw value 0–4095 (firmware-owned)
  waterLevelLow: boolean; // firmware-owned
}

// ---------------------------------------------------------------------------
// Coercion helpers — RTDB values can arrive as unexpected types.
// ---------------------------------------------------------------------------

function asInt(value: unknown, fallback: number): number {
  if (typeof value === 'number' && Number.isFinite(value)) return Math.trunc(value);
  if (typeof value === 'string' && value.trim() !== '') {
    const n = Number(value);
    if (Number.isFinite(n)) return Math.trunc(n);
  }
  return fallback;
}

function asBool(value: unknown, fallback: boolean): boolean {
  if (typeof value === 'boolean') return value;
  return fallback;
}

function asMode(value: unknown): PumpMode {
  return value === 'auto' ? 'auto' : 'manual';
}

// ---------------------------------------------------------------------------
// Parsing / serialisation
// ---------------------------------------------------------------------------

/**
 * Parses a raw RTDB snapshot value into a {@link TowerState}, using
 * null-coalescing defaults for every field (default mode is 'manual', speed 0,
 * state false, intervals 1).
 */
export function parseTowerState(raw: unknown): TowerState {
  const data = (raw && typeof raw === 'object' ? raw : {}) as Record<string, unknown>;
  const sensors =
    data.sensors && typeof data.sensors === 'object'
      ? (data.sensors as Record<string, unknown>)
      : {};

  return {
    pumpSpeed: asInt(data.pump_speed, 0),
    pumpMode: asMode(data.pump_mode),
    pumpState: asBool(data.pump_state, false),
    pumpSwitch: asBool(data.pump_switch, false),
    intervalOnMin: asInt(data.interval_on_min, 1),
    intervalOffMin: asInt(data.interval_off_min, 1),
    moisture: asInt(sensors.moisture, 0),
    waterLevelLow: asBool(sensors.water_level_low, false),
  };
}

/**
 * Serialises a {@link TowerState} to a plain object matching the Firebase
 * schema, including the nested `sensors` key. Used for test round-trips.
 */
export function toMap(state: TowerState): Record<string, unknown> {
  return {
    pump_speed: state.pumpSpeed,
    pump_mode: state.pumpMode,
    pump_state: state.pumpState,
    pump_switch: state.pumpSwitch,
    interval_on_min: state.intervalOnMin,
    interval_off_min: state.intervalOffMin,
    sensors: {
      moisture: state.moisture,
      water_level_low: state.waterLevelLow,
    },
  };
}

// ---------------------------------------------------------------------------
// Computed helpers
// ---------------------------------------------------------------------------

/** Maps a 0.0–1.0 slider percentage to an integer pump speed in 0–255. */
export function sliderToSpeed(percent: number): number {
  const scaled = Math.round(percent * 255);
  return Math.min(255, Math.max(0, scaled));
}

/** Maps {@link TowerState.pumpSpeed} (0–255) to a 0.0–1.0 slider value. */
export function speedPercent(state: TowerState): number {
  return state.pumpSpeed / 255;
}

/** Maps a water-level-low boolean to its display label. */
export function waterLevelLabel(low: boolean): 'Good' | 'Low' {
  return low ? 'Low' : 'Good';
}

/** Returns `'Low'` when the reservoir is low, otherwise `'Good'`. */
export function waterLevelDisplay(state: TowerState): 'Good' | 'Low' {
  return waterLevelLabel(state.waterLevelLow);
}
